### DISCUSSION TITLE
Should I revoke extraneous login tokens?

### DISCUSSION THREAD

**Discussion by Xevion** [fastapi-users/fastapi-users#1076] [Q&A]
I am using FastAPI-Users with SQLAlchemy and a Redis/Cookie strategy/transport for my backend. It is slightly modified to handle dynamic non-standard lifetimes as well as expose the expiration to the API user but is otherwise unchanged.

### Should I revoke extraneous login tokens?

While playing with the API, I noticed that when logging in, previous tokens are not revoked. This feels wrong, to have a token floating around that may or may not be used. I can hit the endpoint a million times and unused tokens will fill my Redis "database".

- Is this a concern I should worry about? Apparently, this is the default behavior, and I'm not really knowledgeable about the security implications this has.
- Should I incur the penalty at login to search the Redis "database" and delete all keys?
    - I could also simply modify the strategy to only allow 1 token (keyed by the user ID) and check like that. It shouldn't be too difficult, modifying it for my own purposes today was pretty easy.

At the minimum, I feel I should be implementing a logout method to log out ALL devices (delete all tokens). But at that point, I'm operating a much more complex system and be doing more with my authentication.

**Comment by juyanghsu** [comment]
I ran into the same problem as you and was asking my colleagues about it. 

What they typically do is that on login, you can send another request to establish a Websocket that ping/pong the user. 

On connection close(or after a period of time with inactivity), you can make a request to logout the user so that the person's token is revoked in Redis, that way you won't have to search for all tokens by the user on login again. 

Since the person's token is revoked, there won't be a need to destroy token on your browser either.

This runs into another problem that I was asking earlier though, which is how does someone validate a Websocket connection with authentication on this library. #1074

**Reply by juyanghsu** [reply]
Workaround posted if you want to have a look

**Reply by Xevion** [reply]
To be honest though, I really don't want to be using WebSockets unless I really _need_ to. Additionally, I kinda wanna let users setup 'Remember Me' for long periods, such that they don't expire when the usage session ends. I personally hate websites that ALWAYS require a re-login, and I don't want my website to be like that.

Might seem a little weird to want long-term tokens but also only one of them, but that's how I want it, personally.

**Reply by Xevion** [reply]
But thanks for the workaround. Since I might need websockets soon depending on the implementation I choose, it's good to get a authentication method ready. Will keep in mind.

**Reply by juyanghsu** [reply]
If it’s remember me then you can probably add a condition to not delete them, but yeah, I think there’s a lot of ways to approach this and whatever you do be sure to share it 😎

**Reply by Xevion** [reply]
For remember me, I am thinking of making a token that stays alive for 7-31 days depending on the type of account/settings. Might be too complex though and I shouldn't be focusing on this part of my application.

**Comment by frankie567** [comment]
Indeed, the previous token is not revoked in the Redis implementation. Indeed, if a user is logged in on a first device, we don't want them to be logged out when logged in to a new one.

However, notice that the [`ex` parameter is set](https://github.com/fastapi-users/fastapi-users/blob/3e44abdbeffbe529120f96a91e1f4499202bbe2f/fastapi_users/authentication/strategy/redis.py#L42) when saving the key into Redis. It means that the key will be automatically deleted by Redis after its expiration (ref: https://redis.io/commands/set/).

Now, if you want a fine control about the active tokens, like showing them to the user and have a way to revoke them, I would suggest to go for the [Database strategy](https://fastapi-users.github.io/fastapi-users/10.1/configuration/authentication/#database) instead. By having the tokens in the database, you have a fine control about how they're structured: you can extend the model with more columns (e.g. a device identifier), build SQL requests and use them to define API routes to list or delete them.

**Reply by Xevion** [reply]
Yeah, when I modified the implementation I found the `ex` parameter and modified it to use the `exat` parameter instead (expire at, instead of expire [in]).

Maybe I could go for a hybrid strategy - for basic checks, Redis would be the fast 'cache' of tokens (all authorized endpints), but for logins, which can be slower as they are not common, they would hold the true token and would call Redis to revoke tokens should the database already have one. Might be complex, but doable.

**Reply by frankie567** [reply]
That could work, yes. But I'm not sure it's worth the hassle: SQL databases are also very fast, so you won't really notice a big performance gap.
